grandpa: pluggable voting rules

andresilva opened the pull request:
This PR adds support for plugging arbitrary voting rules to restrict GRANDPA votes.  A `VotingRule` trait is introduced that allows restricting a GRANDPA vote based on what was the initial best finality target and the current finality target.

The GRANDPA environment uses `SelectChain` to find an initial finality target candidate, additionally the GRANDPA environment restricts the vote to be on 3/4 of the unfinalized chain and furthermore the vote is limited by any pending authority set change. After these mandatory voting rules, the environment uses the given `VotingRule` to further restrict the vote if necessary. 

The node is updated to use a voting rule to always vote behind the best block. I also added a composite voting rule and builder that allows combining multiple voting rules. This might be useful in polkadot to implement voting rules that take into account data availability.

rphmeier commented:
Could the 3/4s rule be repurposed as a voting rule as well?

HCastano commented on core/finality-grandpa/src/voting_rule.rs:
```diff
@@ -0,0 +1,181 @@
+		best_target: &Block::Header,
+		current_target: &Block::Header,
+	) -> Option<(Block::Hash, NumberFor<Block>)> {
+		let restricted_target = self.rules.iter().fold(
```
nit: I think this section is easier to read after a little bit of Rustfmt-ing:

```Rust
        let restricted_target =
            self.rules
                .iter()
                .fold(current_target.clone(), |current_target, rule| {
                    rule.restrict_vote(best_target, &current_target)
                        .and_then(|(hash, _)| self.backend.header(BlockId::Hash(hash)).ok())
                        .and_then(std::convert::identity)
                        .unwrap_or(current_target)
                });
```

andresilva replied:
I don't have any strong opinions about code formatting but this is not how our _undefined_ code style looks like.

andresilva commented:
Going to rejig the PR to instead do this with `SelectChain` rather than creating a new `VotingRule` trait.

andresilva commented on core/finality-grandpa/src/environment.rs:
```diff
@@ -396,16 +398,17 @@ impl<B, E, Block: BlockT, N: Network<Block>, RA, SC> Environment<B, E, Block, N,
 	N: Network<Block> + 'static,
 	N::In: 'static,
 	SC: SelectChain<Block> + 'static,
+	VR: VotingRule<Block>,
```
We need to guarantee it is `Clone` somewhere else, with a trait object we can't enforce that bound.

andresilva commented:
I think that `SelectChain` is not appropriate for implementing voting rules, specifically it is not very amenable to composition. Still the separation seems appropriate to me: `SelectChain` is concerned with picking some "best" chain (longest, GHOST, etc.) which is used as an initial target for finality voting, and then on top of that we optionally restrict it with GRANDPA-specific `VotingRule`s.

I've addressed all the previous comments and moved the 3/4 unfinalized chain voting logic to a `VotingRule`. The only rule which is implicit/mandatory is limiting votes to authority set changes which seems sensible to me since it is fundamental for the security of the protocol itself.

DemiMarie-parity commented on core/finality-grandpa/src/environment.rs:
```diff
@@ -462,7 +462,11 @@ where
 						if *target_header.number() < target_number {
-							unreachable!();
+							unreachable!(
+								"we are traversing backwards from a know block; \
```
typo: `know` → `known`

andresilva replied:
Weird thing, I had the exact same message elsewhere but without the typo. Does this mean I did not copy-paste it when writing the code? :thinking::smile:

rphmeier commented on core/finality-grandpa/src/voting_rule.rs:
```diff
@@ -0,0 +1,270 @@
+	/// returned value **must** be an ancestor of the given `current_target`,
+	/// this also means that a variant must be maintained throughout the
+	/// execution of voting rules wherein `current_target <= best_target`.
+	fn restrict_vote(
```
the generic parameter could also be on the function here

andresilva replied:
We use `VotingRule` as a trait object so we can't have a generic here.

rphmeier commented on core/finality-grandpa/src/voting_rule.rs:
```diff
@@ -0,0 +1,270 @@
+	/// Restrict the given `current_target` vote, returning the block hash and
+	/// number of the block to vote on, and `None` in case the vote should not
+	/// be restricted. `base` is the block that we're basing our votes on in
+	/// order to pick our target (e.g. last finalized block), and `best_target`
```
base should be the last-round-estimate, not last finalized?

rphmeier commented on core/finality-grandpa/src/environment.rs:
```diff
@@ -429,7 +433,7 @@ where
 
 		match self.select_chain.finality_target(block, None) {
-			Ok(Some(mut best_hash)) => {
+			Ok(Some(best_hash)) => {
```
I guess in this context I don't understand the purpose of `finality_target` anymore.

andresilva replied:
I agree, we already had a discussion about `SelectChain` on https://github.com/paritytech/substrate/pull/2869#issuecomment-502267671. Since we now have voting rules to implement the vote target logic what do you think about renaming `finality_target` in `SelectChain` to `best_containing`? In this case `SelectChain` is only responsible for implementing some custom notion of "best".

gavofyork commented:
@rphmeier still ok with this?
